src/PhpUnit/UriInterfaceTestsTrait.php

```diff
@@ -0,0 +1,472 @@
+     * @expectedException InvalidArgumentException
+     */
+    public function testWithSchemeFailed($scheme)
+    {
+        $this->createDefaultUri()->withScheme($scheme);
+    }
+
+    public function withSchemeFailedProvider()
+    {
+        return [
+            'invalid char' => ['in,valid'],
+            'integer like string' => ['123'],
+            'float' => [1.2],
+            'array' => [['foo']],
+            'unknown scheme' => ['yolo'],
+        ];
+    }
+
+    /**
+     * @dataProvider withUserInfoFailedProvider
+     * @expectedException InvalidArgumentException
+     */
+    public function testWithUserInfoFailed($user, $pass)
+    {
+        $this->createDefaultUri()->withUserInfo($user, $pass);
+    }
+
+    public function withUserInfoFailedProvider()
+    {
+        return [
+            'invalid character in user :' => ['igo:r', 'rAsMuZeN'],
+            'invalid character in user @' => ['igo@r', 'rAsMuZeN'],
+            'invalid character in pass' => ['iGoR', 'rasmu@sen'],
+            'array in user' => [['iGoR'], 'rAsMuZeN'],
+            'array in pass' => ['iGoR', ['rAsMuZeN']],
+        ];
+    }
+
+    /**
+     * @dataProvider withHostFailedProvider
+     * @expectedException InvalidArgumentException
+     */
+    public function testWithHostFailed($host)
+    {
+        $this->createDefaultUri()->withHost($host);
+    }
+
+    public function withHostFailedProvider()
+    {
+        return [
+            'dot in front' => ['.example.com'],
+            'hyphen suffix' => ['host.com-'],
+            'multiple dot' => ['.......'],
+            'one dot' => ['.'],
+            'empty label' => ['tot.    .coucou.com'],
+            'space in the label' => ['re view'],
+            'underscore in label' => ['_bad.host.com'],
+            'label too long' => [implode('', array_fill(0, 12, 'banana')) . '.secure.example.com'],
+            'too many labels' => [implode('.', array_fill(0, 128, 'a'))],
+            'Invalid IP format' => ['[127.0.0.1]'],
+            'Invalid IPv6 format' => ['[[::1]]'],
+            'Incomplete IPv6 format' => ['[::1'],
+            'space character in starting label' => ['example. com'],
+            'invalid character in host label' => ["examp\0le.com"],
+            'invalid IP with scope' => ['[127.2.0.1%253]'],
+            'invalid scope IPv6' => ['[ab23::1234%251]'],
+            'invalid scope ID' => ['[fe80::1234%25?@]'],
+            'invalid scope ID with utf8 character' => ['[fe80::1234%25€]'],
+        ];
+    }
+
+    /**
+     * @dataProvider withPortFailedProvider
+     * @expectedException InvalidArgumentException
+     */
+    public function testWithPortFailed($port)
+    {
+        $this->createDefaultUri()->withPort($port);
+    }
+
+    public function withPortFailedProvider()
+    {
+        return [
+            'string' => ['toto'],
+            'invalid port number too low' => ['-23'],
+            'invalid port number too high' => ['10000000'],
```

Comment:
For test range values there is a test technique called boundary testing. https://en.wikipedia.org/wiki/Boundary_testing

This technique helps you to determinate what values to test. For port range 0 and 65536.

You may to test a negative value if you want to. The negative value could be (-1) because is the negative version of the minimum port value (1)